[14.0][ADD] Verifactu: Backport from V16

ljimenezsidoo opened the pull request:
l10n_es_verifactu_oca
- Dependencias en l10n_es_extra_data y l10n_es_irnr
- Aeat.mixin, posicion fiscal y campos del contacto
- Reparadas referencias de las plantillas

De momento el modulo esta instalable y enviando las facturas

@flachica @HaraldPanten

https://github.com/OCA/l10n-spain/issues/4368

Commit:
[ADD] l10n_es_verifactu_oca: Comunicación VERI*FACTU

Envío de los registros de facturación a la sede electrónica de la
Agencia Tributaria en el momento de su producción

AndreuOForgeFlow committed:
[IMP] l10n_es_aeat_verifactu_oca: Activar Journal Hash para compañía

almumu committed:
[IMP] l10n_es_verifactu_oca: Add verifactu hash code + send to AEAT

ozono committed:
[ADD] Primera version QR Verifactu (#62)

ozono committed:
[IMP] l10n_es_verifactu_oca: Primera version QR VERI*FACTU

aritzolea committed:
[IMP] l10n_es_aeat_verifactu: Fixes on views, hash calculated on invoice post and refactor + tests

aritzolea committed:
[IMP] l10n_es_aeat_verifactu_oca: Fixes on views, hash calculated on invoice post and refactor + tests

almumu committed:
[IMP] l10n_es_verifactu_oca: Renaming + improvements + fixes

- Modificación de algunos métodos: control de encadenamiento, registro
  anterior y siguiente.
- Envío por cron.triggers
- Renombrado módulo para quitar lo de 'aeat', y también algunos modelos.
- Control de excepciones al hacer el post de la factura para que
  compruebe que la configuración es correcta.
- Algunas mejoras en el código.

flachicasidoo committed:
[IMP] l10n_es_verifactu_oca: README + journal enable flag

flachicasidoo committed:
[IMP] l10n_es_verifactu: README + journal enable flag

ProcessControl-JLR committed:
[IMP+FIX] l10n_es_verifactu_oca: Registration date + start date + search

- Añade la configuración de claves de registro y claves de impuestos a las plantillas de posiciones fiscales de la localización
- Añade la fecha de inicio para habilitar verifactu en las facturas.
- Corrige la búsqueda de empresas españolas y añade rollback en caso de
  error al enviar documentos.

ProcessControl-JLR committed:
[IMP+FIX] l10n_es_verifactu: Registration date + start date + search

- Añade la configuración de claves de registro y claves de impuestos a las plantillas de posiciones fiscales de la localización
- Añade la fecha de inicio para habilitar verifactu en las facturas.
- Corrige la búsqueda de empresas españolas y añade rollback en caso de
  error al enviar documentos.

almumu committed:
[FIX+IMP] l10n_es_verifactu_oca: Taxes + general improvements

- Añadido algunos mapeos de impuestos que faltaban.
- Controlado que no se puedan validar facturas de venta si no están
  todos sus impuestos mapeados.
- Arreglado algunos errores de configuración y funcionamiento.
- Los impuestos no incluidos en el total no se tenían en cuenta y se
  enviaba un importe erróneo.
- El total impuestos y total factura en verifactu aparecen tanto en
  detalle que se envía, como en el hash como en el qr que se imprime y
  la url de verificación de factura, arreglado en todos los procesos
  afectados
- Revisados los casos de envío en que devuelve "aceptado con errores"
  para enviarlo como subsanación e incidencia.
- Actualizado roadmap y readme con lo que queda pendiente.
- Arreglado tests que fallaban

JordiBForgeFlow committed:
[REF+IMP] l10n_es_verifactu_oca: verifactu.send.* + mocked tests

- Refactor para enviar las facturas desde verifactu.send.queue y guardar las respuestas en verifactu.send.respose
- Improve tests related with the chaining of invoices
- Mandar incidencia en cabecera cuando los registros se han generado pasados 260s del envio a verifactu
- Add tests with mocked responses from verifactu

almumu committed:
[FIX] l10n_es_verifactu_oca: añadido compatibilidad y herencia que faltaba con aeat.mixin

MeritxellAForgeFlow committed:
[IMP] l10n_es_verifactu_oca: Add verifactu send response

ljsalvatierra-factorlibre committed:
[IMP] l10n_es_verifactu_oca: new chaining model

ljsalvatierra-factorlibre committed:
[REF] l10n_es_verifactu_oca: integrate new verifactu invoice model with batch queue sending

almumu committed:
[REF+FIX+IMP] l10n_es_verifactu_oca: Models unification + fixes + imps

- Unificado modelos de verifactu.send.queue con el
  verifactu.invoice.entry que es el encadenamiento
- Refactor y arreglo de test.
- Añadido también algunas mejoras en cuanto a comprobaciones y
  trazabilidad.
- Arreglos y mejoras en el código según revisión del PR.

ozono committed:
[FIX] l10n_es_verifactu_oca: 2 decimals in URL

Formateo importe a dos decimales en URL QR para cumplir límite AEAT y
evitar errores con números flotantes.

pedrobaeza committed:
[IMP] l10n_es_aeat_verifactu_oca: Final adjustments

- Final renaming to l10n_es_verifactu_oca.
- Homogeneous code styling without empyt line inside methods.
- Code compactation friendly to ruff.
- Code optimization.
- A little of tests reorganization.
- Labelling everywhere with "VERI*FACTU" naming according the official
  naming and case: https://sede.agenciatributaria.gob.es/Sede/iva/sistemas-informaticos-facturacion-verifactu.html
- FIX: Send IndicadorMultiplesOT = S only when multiple companies have
  enabled VeriFactu, not when they have Spain as country.
- Proper models naming (without plurals) + descriptions.
- Field names and form labels rationalized with the following criteria:
  * For field technical names, all are prefixed with `verifactu_`.
  * For field names (labels), all are prefixed with `VERI*FACTU ` for
    being all grouped when doing a filter.
  * For form labels, the prefix is explicitly removed for not
    overloading the UI and not repeating information.
- Rename verifactu.invoice.entry.response~datetime to date_response for
  not collapsing with datetime library.
- Rellocate some fields of account.move into verifactu.mixing for being
  available to all the VERI*FACTU documents.
- Fiscal position VERI*FACTU settings moved to a page in the notebook
  for better visual ergonomy.
- Invoice VERI*FACTU page layout restructured for better ergonomy.
- Show all the VERI*FACTU checks errors at once and without repeating
  strings, with a huge refactorization and also contemplating other
  possible tax agencies.
- Remove check about requiring a fiscal position in the invoice. The
  usual case for national customers is to not set "Régimen Nacional", so
  now forcing to set it will increase the entry barrier.
- Translations to Spanish.

Commit:
[UPD] Update l10n_es_verifactu_oca.pot

OCA-git-bot committed:
[BOT] post-merge updates

pedrobaeza committed:
[FIX+IMP]: l10n_es_verifactu_oca: Init time + fine-tuning

- Use hooks for reducing installation time, pre-creating columns with
  some sane defaults (but not accurate).
- Mark as enabled when no fiscal position is set, but the rest of
  condition are OK.
- Don't show VERI*FACTU refund types on vendor bill refunds or if no
  VERI*FACTU enabled.

pedrobaeza committed:
[FIX] l10n_es_verifactu_oca: zeep responses without get

zeep return objects for the responses with no get method.

```
AttributeError: RespuestaRegFactuSistemaFacturacionType instance has no attribute 'get'
````

It wasn't detected in tests because of the mocked objects.

jonpasquier committed:
Added translation using Weblate (Catalán (ca_ES) (ca_ES))

jonpasquier committed:
Added translation using Weblate (Catalan)

jonpasquier committed:
Translated using Weblate (Catalan)

Currently translated at 100.0% (256 of 256 strings)

Translation: l10n-spain-16.0/l10n-spain-16.0-l10n_es_verifactu_oca
Translate-URL: https://translation.odoo-community.org/projects/l10n-spain-16-0/l10n-spain-16-0-l10n_es_verifactu_oca/ca/

ljimenezsidoo committed:
[14.0][MIG] Backport Verifactu OCA

FIX move all to verifactu module

FIX test

FIX coverage

FIX template errors

FIX report address and QR

FIX requested changes

mtbochaca committed:
Translated using Weblate (Catalan)

Currently translated at 100.0% (256 of 256 strings)

Translation: l10n-spain-16.0/l10n-spain-16.0-l10n_es_verifactu_oca
Translate-URL: https://translation.odoo-community.org/projects/l10n-spain-16-0/l10n-spain-16-0-l10n_es_verifactu_oca/ca/

ljimenezsidoo commented:
https://github.com/OCA/l10n-spain/pull/4373 Antiguo PR con algunos comentarios
Este PR tiene los commits de v16 también.

flachica commented:
@HaraldPanten @anmarmo1 @acysos 

Informaros que hemos hecho pruebas funcionales con éxito en esta rama. Queda pendinte pelearnos un poco con los test y ponerla en fase de revisión

Por si estimáis oportuno ese vistazo funcional

Commit:
[IMP] l10n_es_verifactu_oca: Added a QR on the header acording the AEAT requirements.

anmarmo1 reviewed (approved):
A nivel funcional es correcto, me he dado cuenta que me pide el tax key, que no esta configurado en el data, diria que en v.16 pusimos este valor en un data pero parece que no es así @almumu ¿me confirmas si lo pusimos o no?

pedrobaeza commented:
Sobre los tax key, lo llevo pensando un tiempo, y deberíamos evitar los modelos nuevos y asociarlos a los impuestos directamente, ya que no ocurre como en el SII que salen excepciones. ¿Qué opináis?

anmarmo1 commented:
> Sobre los tax key, lo llevo pensando un tiempo, y deberíamos evitar los modelos nuevos y asociarlos a los impuestos directamente, ya que no ocurre como en el SII que salen excepciones. ¿Qué opináis?

Pero entonces habría que controlar que no se mezclasen impuestos o quedarnos con la primera¿?. A ver, teóricamente no se pueden mezclar impuestos ¿no? Ponerlo en los impuestos lo simplificaría a nivel de modelos y datas...

pedrobaeza commented:
> Pero entonces habría que controlar que no se mezclasen impuestos o quedarnos con la primera¿?. A ver, teóricamente no se pueden mezclar impuestos ¿no? Ponerlo en los impuestos lo simplificaría a nivel de modelos y datas...

Ya existe un control para mezcla de impuestos, que puede servir para esto mismo.

almumu committed:
[IMP] Movido QR a la parte de las direcciones de la factura, para que salga al principio y a la izquierda

ljimenezsidoo commented:
@anmarmo1 @pedrobaeza 

Me puedes pasar pantallazo de lo que comentas sobre los tax key

En caso de que queráis que cambie cualquier cosa indicar exactamente el cambio por favor y lo dejo hecho cuanto antes

Gracias por las pruebas y comentarios

almumu committed:
[IMP] code refactor

pedrobaeza committed:
[FIX] l10n_es_verifactu_oca: Proper identifier for receiver

Return the identifier with the corresponding data.

Fixes #4389

pedrobaeza commented:
Finalmente después de analizarlo, creo que no merece cambiar el sistema actual, porque sigue siendo más flexible asociado a las posiciones fiscales, así que con eso vamos a seguir igual. Lo que sí que habría que hacer cherry-pick aquí es de #4398.

HaraldPanten commented:
FYI: A tener en cuenta https://github.com/OCA/l10n-spain/pull/4384

ljimenezsidoo commented:
@pedrobaeza @HaraldPanten 

Se ha hecho cherry-pick del PR https://github.com/OCA/l10n-spain/pull/4398

¿Hace falta hacer lo mismo para el PR https://github.com/OCA/l10n-spain/pull/4384

Gracias!!

HaraldPanten commented:
> @pedrobaeza @HaraldPanten
> 
> Se ha hecho cherry-pick del PR #4398
> 
> ¿Hace falta hacer lo mismo para el PR #4384
> 
> Gracias!!

Hasta donde yo sé, sí. Es necesario que el QR esté en la parte superior, como se ha hecho en la V16.

jdani1296 reviewed (approved):
Probado en v14 funcionalidad y posición qr igual que en v16

flachica reviewed (approved):
LGTM

syci committed:
[16.0][FIX] l10n_es_verifactu_oca: ref err de usuario en mail.activity.type

- No cambia la lógica de envío Veri*Factu ni el contenido del XML remitido.
- Cumple con RD 1007/2023 (integridad y trazabilidad): el ajuste es interno
  de Odoo y no afecta a hashes, QR ni datos fiscales transmitidos.

Este cambio es puramente técnico para garantizar la correcta instalación/
operación del módulo. No modifica el comportamiento del SIF en la
remisión de registros a la AEAT ni su conformidad normativa

pedrobaeza commented:
¿Podeís por favor arreglar el historial de commits?

- Los commits hechos cherry-pick colocarlos delante de vuestro backport.
- Juntar/reordenar los commits que corresponden al backport en si (por autores/set lógicos de cambios).

Habría que tener también en cuenta #4440, aunque eso luego si no, podría yo hacer backport.

ljimenezsidoo commented:
Buenas @pedrobaeza 

He ordenado los commits y añadido el que faltaba también

pedrobaeza commented on l10n_es_verifactu_oca/__manifest__.py:
```diff
@@ -18,7 +18,6 @@
         "l10n_es_aeat",
         "l10n_es_irnr",
         "l10n_es_extra_data",
-        "report_qweb_element_page_visibility",
```
Esto no debería salir en este commit.

ljimenezsidoo replied:
Este cambio viene de este commit que es un cherry-pick: https://github.com/OCA/l10n-spain/commit/8e0ee5bf8af848bde5757adf7ac109288421d5a8

No entiendo porque no tiene que salir si viene de otro commit. ¿Que tengo que hacer?

pedrobaeza commented on l10n_es_verifactu_oca/data/account_fiscal_position_template_data.xml:
```diff
@@ -106,22 +106,22 @@
     </record>
 
-    <record id="l10n_es.fp_irpfnrnue24" model="account.fiscal.position.template">
+    <record id="l10n_es_irnr.fp_irpfnrnue24" model="account.fiscal.position.template">
```
Hay que quitar estos impuestos y no depender en el manifest de este módulo. Si se necesitan, se tendrá que hacer el módulo puente correspondiente. También parece raro que no esté en el commit de backport la adición de ese módulo en el manifest.

pedrobaeza commented on l10n_es_verifactu_oca/data/verifactu_map_data.xml:
```diff
@@ -12,18 +12,18 @@
                 ref('l10n_es.account_tax_template_s_iva21b'),
                 ref('l10n_es.account_tax_template_s_iva0b'),
-                ref('l10n_es.account_tax_template_s_iva2b'),
+                ref('l10n_es_extra_data.account_tax_template_s_iva2b'),
```
Puedes eliminar toda esta parte, ya que esos impuestos no están vigentes desde el año pasado, y no dependas de `l10n_es_extra_data`.

pedrobaeza commented on l10n_es_aeat/models/aeat_mixin.py:
```diff
@@ -0,0 +1,163 @@
+# Copyright 2021 Tecnativa - João Marques
```
Esto se sigue añadiendo a `l10n_es_aeat`, y no debería ser así.

pedrobaeza reviewed (changes requested):
Creo que sigue habiendo un poco de mezcla de commits.

ivilata commented:
Por si sirve para apoyar la mezcla de este PR, he probado a migrarlo a 15.0 con `oca-port` y, con un par de ajustes menores en los tests y en los datos, también supera los tests y parece funcionar bien, con lo que la migración a 15.0 sería casi inmediata. :+1:

HaraldPanten commented:
> Por si sirve para apoyar la mezcla de este PR, he probado a migrarlo a 15.0 con `oca-port` y, con un par de ajustes menores en los tests y en los datos, también supera los tests y parece funcionar bien, con lo que la migración a 15.0 sería casi inmediata. 👍

¿Podríais hacer el PR en OCA, y así podemos gestionarlo comunitariamente?

Saludos.

ivilata commented:
@HaraldPanten:

> ¿Podríais hacer el PR en OCA, y así podemos gestionarlo comunitariamente?

Estaba esperando a que este PR se cerrara (cosa que parece inminente) para evitar repetir trabajo en la preparación de mi PR, pero he aprovechado igualmente para ir haciendo. Aquí está: #4503 

Hasta que se cierre este PR dejaré el otro en modo boceto, pero si lo veis muy muy bien puedo “ascenderlo” a PR oficial. :wink: (@pedrobaeza)

pedrobaeza reviewed (approved):
/ocabot merge nobump

ivilata commented:
¡Gracias @ljimenezsidoo por el PR y @pedrobaeza por mezclarlo! Ya he pasado #4503 (migración a 15) a PR “completo”. :slightly_smiling_face:
